pkg/build/build.go

```diff
@@ -554,30 +554,110 @@ func WithEnvFile(envFile string) Option {
+// ready to use for config parsing.
+func (options *configOptions) include(opts ...ConfigurationParsingOption) {
+	for _, fn := range opts {
+		fn(options)
+	}
+	if options.logger == nil {
+		options.logger = nopLogger{}
+	}
+}
+
+// open handles opening of a file with the given name, in such a way that
+// automatically chooses to use an abstract or concrete OS filesystem.
+func (options *configOptions) open(filepath string) (fs.File, error) {
+	if filesystem := options.abstractFS; filesystem != nil {
+		return filesystem.Open(filepath)
+	}
+
+	return os.Open(filepath)
+}
+
+// WithAbstractFS adjusts the behavior of configuration parsing to expect all
+// file paths to be relative to the abstract fs.FS. This approach isn't supported
+// yet except for when opening the config file.
+func WithAbstractFS(filesystem fs.FS) ConfigurationParsingOption {
+	return func(options *configOptions) {
+		options.abstractFS = filesystem
 	}
+}
 
-	if err := yaml.Unmarshal(data, cfg); err != nil {
-		return fmt.Errorf("unable to parse configuration file: %w", err)
+// WithConfigFile sets the path from which to read the configuration file.
+func WithConfigFile(path string) ConfigurationParsingOption {
+	return func(options *configOptions) {
+		options.configFilePath = path
 	}
+}
 
+// WithEnvFileForParsing set the paths from whcih to read an environment file.
+func WithEnvFileForParsing(path string) ConfigurationParsingOption {
+	return func(options *configOptions) {
+		options.envFilePath = path
+	}
+}
+
+// WithLogger sets the logger to use during configuration parsing. This is
+// optional, and if not supplied, a no-op logger will be used.
+func WithLogger(logger Logger) ConfigurationParsingOption {
+	return func(options *configOptions) {
+		options.logger = logger
+	}
+}
+
+func detectCommit(options *configOptions) string {
 	// Best-effort detection of current commit, to be used when not specified in the config file
-	detectedCommit := ""
-	repo, err := git.PlainOpen(path.Dir(ctx.ConfigFile))
+
+	// TODO: figure out how to use an abstract FS
+	repo, err := git.PlainOpen(path.Dir(options.configFilePath))
 	if err != nil {
-		ctx.Logger.Printf("unable to detect git commit for build configuration: %v", err)
-	} else {
-		head, err := repo.Head()
-		if err == nil {
-			detectedCommit = head.Hash().String()
-			ctx.Logger.Printf("detected git commit for build configuration: %s", detectedCommit)
-		}
+		options.logger.Printf("unable to detect git commit for build configuration: %v", err)
+		return ""
 	}
 
+	head, err := repo.Head()
+	if err != nil {
+		return ""
+	}
+
+	commit := head.Hash().String()
+	options.logger.Printf("detected git commit for build configuration: %s", commit)
+	return commit
+}
+
+// ParseConfiguration returns a decoded build Configuration using the parsing options provided.
+func ParseConfiguration(opts ...ConfigurationParsingOption) (*Configuration, error) {
```

Comment:
Maybe the config file itself should be a normal parameter since we'll always want to have it 🤔

Reply:
👍